documentation/general/workloads/workload-history.md

```diff
@@ -0,0 +1,44 @@
+# Workload History
+
+When .NET SDK workloads are updated, it is possible that the new versions of the workloads won't work (if we shipped a bug) or won't be available (if the manifest updates are available but the packs themselves are not).  Right now it is hard to fix such a state.  We will add a `dotnet workload history` command that will show a history of the workload installation actions that have been run via the .NET CLI, and add new options on the `dotnet workload update` command that will allow rolling back to a previously installed version of workloads from the history.
+
+## Recording installation actions
+
+When a CLI command is run which modifies the installed workloads or updates the manifests, we will write a record to the workload history log.  The workload commands this applies to include `install`, `update`, `uninstall`, `repair`, and `restore`.  Each workload history record will include the following information:
+
+- Date/time command was run
+- Date/time command completed
+- Top-level workload command run (such as install, update, or uninstall)
+- Workloads passed as arguments
+- Contents of rollback file passed in, if any
+- Full command line
+- Success or failure, and error message on failure
+- Workload state before and after command run
+  - Manifest versions (rollback file)
+  - Installed workloads
+
+Each installation record will be stored as a file in a folder specific to the SDK feature band.  For MSI-based installs, this folder will be `%PROGRAMDATA%\dotnet\workloads\{sdk-band}\history\`.  For File-based installs, it will be `{DOTNET ROOT}\metadata\workloads\{sdk-band}\history\`.
+
+## `dotnet workload history`
+
+The `dotnet workload history` command will show the history of workload installation commands.  The output will be similar to the following format:
+
+|ID|Date|Command|Workloads|
+|--|----|-------|---------|
+|3|2023-02-01|Update|android, maui, wasm-tools|
+| |2023-02-01|Unknown|Unknown|
+|2|2023-01-05|Install|maui, wasm-tools|
+|1|2023-01-01|Install|android|
```

Comment:
Based on our discussion yesterday, I think we're converging on a design.  Here's the updates I would suggest:

- Reverse the order of the workload history so that the latest changes are listed last (it looks like the implementation is already this way, so just this doc needs to be updated).
- Add an "initial state" row with ID 1 that represents the state before the first history record.  This means that `dotnet workload update --from-history 1` will generally roll back to the baseline manifests.
- Add an ID to the "Unknown" rows.  This will allow rolling back from workload commands even if the previous action was "unknown"
- Try to reword "Unknown" to be more understandable.  Perhaps "Unlogged changes"?  Also, if there are any of these rows, add a note at the bottom explaining something like "Rows with Unlogged changes represent changes from actions other than .NET CLI workload commands.  Usually this represents an update to the .NET SDK or to Visual Studio."
- Once we add support for workload sets, we should add a workload version column to the table.  That doesn't need to be in the initial PR we merge for workload history though, we can add it later.

@baronfel What do you think of these proposals?  Any suggestions for better wording?

Reply:
Those all sound reasonable to me. I hadn't even noticed that the ordering was opposite to what was in the design. Might've just had that wrong in my head after reading the original spec.

Reply:
Definitely agree on all points, and I can help brainstorm a name for that concept